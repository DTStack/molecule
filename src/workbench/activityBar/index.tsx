import 'reflect-metadata';
import { connect } from 'mo/react';
import { ActivityBarController } from 'mo/controller';

import ActivityBar from './activityBar';
import { container } from 'tsyringe';
import { ActivityBarService } from 'mo/services';
export * from './activityBar';
export { ActivityBarItem } from './activityBarItem';

const activityBarService = container.resolve(ActivityBarService);
const activityBarController = container.resolve(ActivityBarController);

export const ActivityBarView = connect(
    activityBarService,
    ActivityBar,
    activityBarController
);
