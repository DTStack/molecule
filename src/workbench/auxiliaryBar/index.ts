import 'reflect-metadata';
import { connect } from 'mo/react';
import { container } from 'tsyringe';
import { AuxiliaryController } from 'mo/controller';
import { AuxiliaryBarService } from 'mo/services';
import { default as AuxiliaryBarView } from './auxiliaryBar';
import { default as AuxiliaryBarTabView } from './auxiliaryBarTab';

const auxiliaryService = container.resolve(AuxiliaryBarService);
const auxiliaryController = container.resolve(AuxiliaryController);

const AuxiliaryBar = connect(auxiliaryService, AuxiliaryBarView);

const AuxiliaryBarTab = connect(
    auxiliaryService,
    AuxiliaryBarTabView,
    auxiliaryController
);

export { AuxiliaryBar, AuxiliaryBarTab };
