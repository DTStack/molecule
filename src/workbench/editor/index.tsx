import 'reflect-metadata';
import { connect } from 'mo/react';
import { container } from 'tsyringe';
import { EditorController } from 'mo/controller/editor';
import { Editor } from './editor';
import { EditorService, LayoutService } from 'mo/services';

const editorService = container.resolve(EditorService);
const layoutService = container.resolve(LayoutService);
import { EditorStatusBarView } from './statusBarView';

const editorController = container.resolve(EditorController);

const EditorView = connect(
    { editor: editorService, layout: layoutService },
    Editor,
    editorController
);

export { Editor, EditorStatusBarView, EditorView };
