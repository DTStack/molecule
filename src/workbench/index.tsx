export { WorkbenchView as Workbench } from './workbench';
export { ActivityBar } from 'mo/workbench/activityBar';
export { ActivityBarItem } from 'mo/workbench/activityBar/activityBarItem';
export { Editor } from 'mo/workbench/editor/editor';
export { EditorGroup } from 'mo/workbench/editor/group';
export { MenuBar } from 'mo/workbench/menuBar/menuBar';
export { Panel } from 'mo/workbench/panel/panel';
export { Sidebar } from 'mo/workbench/sidebar';
export { StatusBar, StatusItem } from 'mo/workbench/statusBar';

export type {
    IWorkbench,
    IActivityBar,
    IActivityBarItem,
    IEditor,
    IEditorGroup,
    IMenuBar,
    IPanel,
    ISidebar,
    IStatusBar,
    IStatusBarItem,
    IProblems,
    IProblemsItem,
    INotification,
    INotificationItem,
} from 'mo/model';

export type {
    IActivityBarController,
    ILayoutController,
    IEditorController,
    ISideBarController,
    IMenuBarController,
    ISettingsController,
    INotificationController,
    IPanelController,
    IProblemsController,
    IStatusBarController,
} from 'mo/controller';
