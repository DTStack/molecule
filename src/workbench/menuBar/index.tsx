import 'reflect-metadata';
import { connect } from 'mo/react';

import MenuBar from './menuBar';
import { MenuBarService } from 'mo/services';
import { container } from 'tsyringe';
import { MenuBarController } from 'mo/controller/menuBar';

const menuBarService = container.resolve(MenuBarService);
const menuBarController = container.resolve(MenuBarController);

const MenuBarView = connect(menuBarService, MenuBar, menuBarController);

export { MenuBar, MenuBarView };
