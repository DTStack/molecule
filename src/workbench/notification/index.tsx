import NotificationPane from './notificationPane';
import NotificationStatusBarView from './statusBarView';

export { NotificationPane, NotificationStatusBarView };
