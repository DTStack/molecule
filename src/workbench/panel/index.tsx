import 'reflect-metadata';
import { container } from 'tsyringe';
import { connect } from 'mo/react';
import { IPanelService, PanelService } from 'mo/services';
import Panel from './panel';
import { PanelController } from 'mo/controller/panel';

const panelService = container.resolve<IPanelService>(PanelService);
const panelController = container.resolve(PanelController);

const PanelView = connect(panelService, Panel, panelController);

export { PanelView };
