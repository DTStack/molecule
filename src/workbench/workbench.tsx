import 'reflect-metadata';
import React from 'react';
import { container } from 'tsyringe';

import {
    classNames,
    getFontInMac,
    prefixClaName,
    getBEMModifier,
    getBEMElement,
} from 'mo/common/className';

import { EditorView } from 'mo/workbench/editor';
import { SidebarView } from 'mo/workbench/sidebar';
import { MenuBarView } from 'mo/workbench/menuBar';
import { ActivityBarView } from 'mo/workbench/activityBar';
import { StatusBarView } from 'mo/workbench/statusBar';
import { PanelView } from 'mo/workbench/panel';
import { ID_APP } from 'mo/common/id';
import { APP_PREFIX } from 'mo/common/const';

import { connect } from 'mo/react';

import { ILayoutController, LayoutController } from 'mo/controller/layout';
import { LayoutService } from 'mo/services';
import { ILayout, MenuBarMode } from 'mo/model/workbench/layout';

import { IWorkbench } from 'mo/model';
import SplitPane from 'mo/components/split/SplitPane';
import { Pane } from 'mo/components/split';
import { Display } from 'mo/components';

const mainBenchClassName = prefixClaName('mainBench');
const workbenchClassName = prefixClaName('workbench');
const compositeBarClassName = prefixClaName('compositeBar');
const appClassName = classNames(APP_PREFIX, getFontInMac());
const workbenchWithHorizontalMenuBarClassName = getBEMModifier(
    workbenchClassName,
    'with-horizontal-menuBar'
);
const withHiddenStatusBar = getBEMModifier(
    workbenchClassName,
    'with-hidden-statusBar'
);
const displayActivityBarClassName = getBEMElement(
    workbenchClassName,
    'display-activityBar'
);

const layoutController = container.resolve(LayoutController);
const layoutService = container.resolve(LayoutService);

export function WorkbenchView(props: IWorkbench & ILayout & ILayoutController) {
    const {
        activityBar,
        menuBar,
        panel,
        sidebar,
        statusBar,
        onPaneSizeChange,
        onHorizontalPaneSizeChange,
        splitPanePos,
        horizontalSplitPanePos,
    } = props;

    const getSizes = () => {
        if (panel.hidden) {
            return ['100%', 0];
        }
        if (panel.panelMaximized) {
            return [0, '100%'];
        }
        return horizontalSplitPanePos;
    };

    const isMenuBarVertical =
        !menuBar.hidden && menuBar.mode === MenuBarMode.vertical;
    const isMenuBarHorizontal =
        !menuBar.hidden && menuBar.mode === MenuBarMode.horizontal;
    const horizontalMenuBar = isMenuBarHorizontal
        ? workbenchWithHorizontalMenuBarClassName
        : null;
    const hideStatusBar = statusBar.hidden ? withHiddenStatusBar : null;
    const workbenchFinalClassName = classNames(
        workbenchClassName,
        horizontalMenuBar,
        hideStatusBar
    );

    return (
        <div id={ID_APP} className={appClassName} tabIndex={0}>
            <div className={workbenchFinalClassName}>
                <Display visible={isMenuBarHorizontal}>
                    <MenuBarView mode={MenuBarMode.horizontal} />
                </Display>
                <div className={mainBenchClassName}>
                    <div className={compositeBarClassName}>
                        <Display visible={isMenuBarVertical}>
                            <MenuBarView mode={MenuBarMode.vertical} />
                        </Display>
                        <Display
                            visible={!activityBar.hidden}
                            className={displayActivityBarClassName}
                        >
                            <ActivityBarView />
                        </Display>
                    </div>
                    <SplitPane
                        sizes={sidebar.hidden ? [0, '100%'] : splitPanePos}
                        split="vertical"
                        showSashes={!sidebar.hidden}
                        allowResize={[false]}
                        onChange={onPaneSizeChange!}
                    >
                        <Pane minSize={170} maxSize="80%">
                            <SidebarView />
                        </Pane>
                        <SplitPane
                            sizes={getSizes()}
                            showSashes={!panel.hidden && !panel.panelMaximized}
                            allowResize={[true, false]}
                            split="horizontal"
                            onChange={onHorizontalPaneSizeChange!}
                        >
                            <Pane minSize="10%" maxSize="80%">
                                <EditorView />
                            </Pane>
                            <PanelView />
                        </SplitPane>
                    </SplitPane>
                </div>
            </div>
            <Display visible={!statusBar.hidden}>
                <StatusBarView />
            </Display>
        </div>
    );
}

export const Workbench = connect(
    layoutService,
    WorkbenchView,
    layoutController
);
