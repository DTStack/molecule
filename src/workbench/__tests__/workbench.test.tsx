import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';

import { WorkbenchView, Workbench } from '../workbench';
import {
    ActivityBarModel,
    IActivityBar,
    IMenuBar,
    IPanel,
    ISidebar,
    IStatusBar,
    IWorkbench,
    MenuBarModel,
    PanelModel,
    SidebarModel,
    StatusBarModel,
} from 'mo/model';
import {
    ILayout,
    IPanelViewState,
    ISidebarViewState,
    LayoutModel,
    ViewVisibility,
    IMenuBarViewState,
    MenuBarMode,
} from 'mo/model/workbench/layout';
import { select, selectAll } from 'mo/common/dom';
import {
    sashHorizontalClassName,
    splitClassName,
    paneItemClassName,
} from 'mo/components/split/base';
import { sleep } from '@test/utils';

function expectElementInOrNot(
    ele: Element | null,
    InDocument: boolean,
    horizontal: boolean = true
) {
    if (InDocument) {
        expect(
            ele?.parentElement?.style[horizontal ? 'height' : 'width']
        ).not.toBe('0px');
    } else {
        expect(ele?.parentElement?.style[horizontal ? 'height' : 'width']).toBe(
            '0px'
        );
    }
}

/**
 * Should display the Editor, Panel, Sidebar, ActivityBar, StatusBar, and MenuBar so on Views
 * @param container
 */
function expectBasicPartsInTheDocument() {
    expectElementInOrNot(select('.mo-editor'), true);
    expectElementInOrNot(select('.mo-panel'), true);
    expectElementInOrNot(select('.mo-sidebar'), true);
    expect(select('.mo-activityBar')).toBeInTheDocument();
    expect(select('.mo-statusBar')).toBeInTheDocument();
    expect(select('.mo-menuBar')).toBeInTheDocument();
}

describe('Test Workbench Component', () => {
    let original;
    const observerFnCollection: any[] = [];
    beforeEach(() => {
        original = HTMLElement.prototype.getBoundingClientRect;
        // @ts-ignore
        HTMLElement.prototype.getBoundingClientRect = () => ({
            width: 500,
            height: 500,
        });

        global.ResizeObserver = jest.fn().mockImplementation((fn) => {
            fn();
            observerFnCollection.push(fn);
            return {
                observe: jest.fn(),
                unobserve: jest.fn(),
                disconnect: jest.fn(),
            };
        });
    });

    afterEach(() => {
        HTMLElement.prototype.getBoundingClientRect = original;
        observerFnCollection.length = 0;
    });

    function workbenchModel(): IWorkbench & ILayout {
        const panel = new PanelModel();
        const activityBar = new ActivityBarModel();
        const menuBar = new MenuBarModel();
        const statusBar = new StatusBarModel();
        const sidebar = new SidebarModel();
        const layout = new LayoutModel();

        const panelState = Object.assign<IPanel, IPanelViewState>(
            panel,
            layout.panel
        );
        const activityBarState = Object.assign<IActivityBar, ViewVisibility>(
            activityBar,
            layout.activityBar
        );
        const menuBarState = Object.assign<IMenuBar, IMenuBarViewState>(
            menuBar,
            layout.menuBar
        );
        const statusBarState = Object.assign<IStatusBar, ViewVisibility>(
            statusBar,
            layout.statusBar
        );
        const sidebarState = Object.assign<ISidebar, ISidebarViewState>(
            sidebar,
            layout.sidebar
        );

        return {
            panel: panelState,
            activityBar: activityBarState,
            menuBar: menuBarState,
            statusBar: statusBarState,
            sidebar: sidebarState,
            splitPanePos: layout.splitPanePos,
            horizontalSplitPanePos: layout.horizontalSplitPanePos,
            groupSplitPos: layout.groupSplitPos,
            editorGroupDirection: layout.editorGroupDirection,
        };
    }

    test('Match The WorkbenchView snapshot', () => {
        const component = render(<WorkbenchView {...workbenchModel()} />);
        const tree = component.asFragment();
        expect(tree).toMatchSnapshot();
    });

    test('Workbench should render all basic parts', () => {
        render(<Workbench />);
        expectBasicPartsInTheDocument();
    });

    test('WorkbenchView should render all basic parts', async () => {
        const workbench = workbenchModel();
        render(<WorkbenchView {...workbench} />);
        expectBasicPartsInTheDocument();
    });

    test('Listen to The WorkbenchView onPaneSizeChange event', async () => {
        const fn = jest.fn();
        render(<WorkbenchView {...workbenchModel()} onPaneSizeChange={fn} />);

        const sashs = selectAll<HTMLDivElement>('div[role="Resizer"]');
        const wrapper = select<HTMLDivElement>(`.${splitClassName}`);

        fireEvent.mouseDown(sashs[0], { screenX: 0, screenY: 0 });
        fireEvent.mouseMove(wrapper!, { screenX: 10, screenY: 10 });
        fireEvent.mouseUp(wrapper!);

        expect(fn).toBeCalled();
        // Compare the splitPanePos arguments
        expect(fn.mock.calls[0][0].length).toBe(2);
    });

    test('Listen to The WorkbenchView onHorizontalPaneSizeChange event', async () => {
        const fn = jest.fn();
        render(
            <WorkbenchView
                {...workbenchModel()}
                onHorizontalPaneSizeChange={fn}
            />
        );
        const sashs = selectAll<HTMLDivElement>(`.${sashHorizontalClassName}`);
        const wrapper = selectAll<HTMLDivElement>(`.${splitClassName}`)[1];

        fireEvent.mouseDown(sashs[0], { screenX: 0, screenY: 0 });
        fireEvent.mouseMove(wrapper!, { screenX: 10, screenY: 10 });
        fireEvent.mouseUp(wrapper!);

        expect(fn).toBeCalled();
    });

    test('Hide the Panel view', async () => {
        const workbench = workbenchModel();
        const { rerender } = render(<WorkbenchView {...workbench} />);
        expectElementInOrNot(select('.mo-panel'), true);
        workbench.panel.hidden = true;
        rerender(<WorkbenchView {...workbench} />);
        expectElementInOrNot(select('.mo-panel'), false);
    });

    test('Maximize the Panel', async () => {
        const workbench = workbenchModel();
        const { rerender } = render(<WorkbenchView {...workbench} />);
        expectElementInOrNot(select('.mo-editor'), true);
        workbench.panel.panelMaximized = true;
        rerender(<WorkbenchView {...workbench} />);
        expectElementInOrNot(select('.mo-editor'), false);

        workbench.panel.panelMaximized = false;
        rerender(<WorkbenchView {...workbench} />);
        expectElementInOrNot(select('.mo-editor'), true);
    });

    test('Set the panel hidden and panelMaximized', async () => {
        const workbench = workbenchModel();

        workbench.panel.panelMaximized = true;
        const { rerender } = render(<WorkbenchView {...workbench} />);
        expectElementInOrNot(select('.mo-editor'), false);
        expectElementInOrNot(select('.mo-panel'), true);

        workbench.panel.hidden = true;
        workbench.panel.panelMaximized = true;
        rerender(<WorkbenchView {...workbench} />);
        expectElementInOrNot(select('.mo-editor'), true);
        expectElementInOrNot(select('.mo-panel'), false);
    });

    test('Hide the Sidebar', async () => {
        const workbench = workbenchModel();
        const { rerender } = render(<WorkbenchView {...workbench} />);
        expectElementInOrNot(select('.mo-sidebar'), true, false);
        workbench.sidebar.hidden = true;
        rerender(<WorkbenchView {...workbench} />);

        expectElementInOrNot(select('.mo-sidebar'), false, false);
    });

    test('Hide the StatusBar', async () => {
        const workbench = workbenchModel();
        const { rerender } = render(<WorkbenchView {...workbench} />);
        expect(select('.mo-statusBar')).toBeInTheDocument();
        workbench.statusBar.hidden = true;
        rerender(<WorkbenchView {...workbench} />);
        expect(select('.mo-statusBar')?.parentElement?.style.display).toBe(
            'none'
        );
    });

    test('Should support to change the layout mode of MenuBar', async () => {
        const workbench = workbenchModel();
        workbench.menuBar.mode = MenuBarMode.vertical;
        const { rerender } = render(<WorkbenchView {...workbench} />);
        expect(select('.mo-menuBar')).toBeInTheDocument();

        workbench.menuBar.mode = MenuBarMode.horizontal;
        rerender(<WorkbenchView {...workbench} />);
        expect(select('.mo-menuBar--horizontal')).toBeInTheDocument();
    });

    test('Should resize panes when called on ResizeObserver', async () => {
        const workbench = workbenchModel();
        const horizontalMockFn = jest.fn();
        const paneChangeMockFn = jest.fn();
        const { container } = render(
            <WorkbenchView
                {...workbench}
                onHorizontalPaneSizeChange={horizontalMockFn}
                onPaneSizeChange={paneChangeMockFn}
            />
        );

        await act(async () => {
            // mock resize
            // @ts-ignore
            HTMLElement.prototype.getBoundingClientRect = () => ({
                width: 1000,
                height: 1000,
            });
            observerFnCollection.forEach((f) => f());
            await sleep(150);
        });

        const panes = container.querySelectorAll<HTMLDivElement>(
            `.${paneItemClassName}`
        );

        expect(panes.length).toBe(4);
        expect(panes[0].style.width).toBe('300px');
        expect(panes[1].style.width).toBe('700px');
        expect(panes[2].style.height).toBe('850px');
        expect(panes[3].style.height).toBe('150px');
    });
});
