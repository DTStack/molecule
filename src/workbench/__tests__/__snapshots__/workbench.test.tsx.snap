// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Workbench Component Match The WorkbenchView snapshot 1`] = `
<DocumentFragment>
  <div
    class="mo"
    id="molecule"
    tabindex="0"
  >
    <div
      class="mo-workbench"
    >
      <div
        class="mo-mainBench"
      >
        <div
          class="mo-compositeBar"
        >
          <div
            class="mo-menuBar"
          >
            <div
              class="mo-drop-down mo-drop-down--right mo-menuBar__action"
            >
              <span
                class="codicon codicon-menu"
              />
            </div>
          </div>
          <div
            class="mo-activityBar"
            id="activityBar"
          >
            <div
              class="mo-activityBar__container"
            >
              <ul>
                <li
                  class="mo-activityBar__item"
                  id="sidebar.search.title"
                >
                  <span
                    class="mo-activityBar__label codicon codicon-search"
                    title="搜索"
                  />
                </li>
              </ul>
              <ul
                class="mo-activityBar__global"
              />
            </div>
          </div>
        </div>
        <div
          class="css-1ndrsdj"
          data-split="vertical"
          data-type="SplitPane"
        >
          <div
            style="display: -webkit-box,-moz-box,-ms-flexbox,-webkit-flex,flex; outline: none; min-width: 170px; max-width: calc(80% - 1px*0.8); flex-grow: 0; width: 300px;"
          >
            <div
              class="mo-sidebar"
            >
              <div
                class="mo-sidebar__pane"
                data-id="sidebar.search.title"
                style="display: none;"
              >
                <div
                  class="sidebar-search-pane"
                >
                  <header
                    class="mo-sidebar__header"
                  >
                    <div
                      class="mo-sidebar__title"
                    >
                      <h2>
                        搜索
                      </h2>
                    </div>
                    <div
                      class="mo-sidebar__toolbar"
                    >
                      <div
                        class="mo-tool-bar"
                      >
                        <div
                          class="mo-action-bar"
                        >
                          <ul
                            class="mo-action-bar__container"
                          >
                            <li
                              class="mo-action-bar__item mo-action-bar__item--disabled"
                              id="search.toolbar.refresh"
                            >
                              <span
                                class="codicon codicon-refresh"
                              />
                            </li>
                            <li
                              class="mo-action-bar__item mo-action-bar__item--disabled"
                              id="search.toolbar.clearAll"
                            >
                              <span
                                class="codicon codicon-clear-all"
                              />
                            </li>
                            <li
                              class="mo-action-bar__item mo-action-bar__item--disabled"
                              id="search.toolbar.collapseAll"
                            >
                              <span
                                class="codicon codicon-collapse-all"
                              />
                            </li>
                          </ul>
                        </div>
                      </div>
                    </div>
                  </header>
                  <div
                    class="mo-sidebar__content"
                  >
                    <div
                      class="mo-search"
                    >
                      <span
                        class="mo-search__replace__button codicon codicon-chevron-right"
                      />
                      <div
                        class="mo-search__group"
                      >
                        <div
                          class="mo-search__input mo-search__target"
                        >
                          <textarea
                            autocapitalize="off"
                            autocorrect="off"
                            class=""
                            placeholder="Search"
                            spellcheck="false"
                            title="Search"
                          />
                          <div
                            class="mo-action-bar mo-search__toolbar"
                          >
                            <ul
                              class="mo-action-bar__container"
                            >
                              <li
                                class="mo-action-bar__item"
                                id="search.matchCase"
                              >
                                <span
                                  class="codicon codicon-case-sensitive"
                                />
                              </li>
                              <li
                                class="mo-action-bar__item"
                                id="search.matchWholeWord"
                              >
                                <span
                                  class="codicon codicon-whole-word"
                                />
                              </li>
                              <li
                                class="mo-action-bar__item"
                                id="search.useRegularExpression"
                              >
                                <span
                                  class="codicon codicon-regex"
                                />
                              </li>
                            </ul>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="mo-tree mo-search__tree"
                      draggable="false"
                      role="tree"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-htpNat fXAXjb sc-bdVaJa cjjWdp"
            data-attribute="vertical"
            data-type="Resizer"
          />
          <div
            style="display: -webkit-box,-moz-box,-ms-flexbox,-webkit-flex,flex; outline: none; min-width: 0; max-width: calc(100% - 1px*1); flex: 1;"
          >
            <div
              class="css-f7b5xf"
              data-split="horizontal"
              data-type="SplitPane"
            >
              <div
                style="display: -webkit-box,-moz-box,-ms-flexbox,-webkit-flex,flex; outline: none; min-height: calc(10% - 1px*0.1); max-height: calc(100% - 1px*1); flex-grow: 0; height: calc(70% - 1px*0.7);"
              >
                <div
                  class="mo-editor"
                >
                  <div
                    class="mo-welcome"
                  >
                    <h1>
                      Molecule.
                    </h1>
                  </div>
                </div>
              </div>
              <div
                class="sc-bwzfXH hEACtv sc-bdVaJa cjjWdp"
                data-attribute="horizontal"
                data-type="Resizer"
              />
              <div
                style="display: -webkit-box,-moz-box,-ms-flexbox,-webkit-flex,flex; outline: none; min-height: 0; max-height: calc(100% - 1px*1); flex: 1;"
              >
                <div
                  class="mo-panel"
                >
                  <div
                    class="mo-panel__header"
                  >
                    <div
                      class="mo-tabs mo-tabs--line"
                    >
                      <div
                        class="mo-tabs__header"
                      >
                        <div
                          class="mo-tab__item"
                          draggable="true"
                        >
                          问题
                        </div>
                      </div>
                    </div>
                    <div
                      class="mo-action-bar mo-panel__toolbar"
                    >
                      <ul
                        class="mo-action-bar__container"
                      />
                    </div>
                  </div>
                  <div
                    class="mo-panel__container"
                    tabindex="0"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mo-statusBar"
      id="statusBar"
    >
      <div
        class="mo-statusBar__left-items"
      >
        <div
          class="mo-statusBar__item"
          id="statusbar.problems.title"
          sortindex="1"
        >
          <a
            tabindex="-1"
            title="Problems"
          >
            <span
              class="codicon codicon-error"
            />
             0 
            <span
              class="codicon codicon-warning"
            />
             0 
            <span
              class="codicon codicon-info"
            />
             0
          </a>
        </div>
      </div>
      <div
        class="mo-statusBar__right-items"
      >
        <div
          actionbar="[object Object],[object Object]"
          class="mo-statusBar__item"
          id="MO_NOTIFICATION"
          sortindex="1"
        >
          <a
            tabindex="-1"
            title="Notification"
          >
            <span
              class="codicon codicon-bell"
            />
          </a>
        </div>
        <div
          class="mo-statusBar__item"
          id="MoEditorInfo"
          sortindex="2"
        >
          <a
            tabindex="-1"
            title="Go to Line/Column"
          >
            <span>
              Ln 0, Col 0
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
